### ISSUE TITLE
Suppressing warnings/messages

### ISSUE THREAD

**Issue by communistkiro** [lfos/calcurse#298]
**Version information.** 4.6.0

**Bug description.** To get to the main interactive calcurse screen, I must jackhammer a key four times. 

**Reproduce.** E.g., with https://paste.debian.net/hidden/55688842/, I get told on every startup that this and that key had been assigned to this and that action as per some default, but that I'd already used them elsewhere.

**Expected Behavior.** All actions have set keybindings at the start. If a user has willingly, consciously removed these, must he be reminded that actions he doesn't need were attempted to be bound to keys he uses for other ones?

- The `-q` option should suppress these warnings.
- Possibility to turn off warnings, or have them removed

**Comment by lhca** [comment]
> **Reproduce.** E.g., with https://paste.debian.net/hidden/55688842/, I get told on every startup that this and that key had been assigned to this and that action as per some default, but that I'd already used them elsewhere.

No bug demonstrated.

The problems stem from _your_ specific setup which you describe somewhat carelessly, or rather not at all. The link above is to a calcurse keys configuration file which looks like an entry for an "obfuscated keyboard mapping" contest. How did you end up with that?

**Comment by communistkiro** [comment]
Not a bug per se, but I saw no suggestions.

Given one can edit keybindings, I wouldn't say exchanging two, unmapping others, or having preferred keys are obfuscation. Rather than pointing a finger at me, either remove users' re- and unmapping capabilities, or make it so it works as one would expect- unmapped actions don't tacitly get mapped (to keys you've explicitly already bound to other action).

**Comment by lhca** [comment]
Now we are getting somewhere. Your original complaint wasn't exactly helpful.

> **Bug description.** To get to the main interactive calcurse screen, I must jackhammer a key four times.

I have tried the keys configuration file you linked to. You have to acknowledge three messages, all concerning undefined key mappings.

> Given one can edit keybindings, I wouldn't say exchanging two, unmapping others, or having preferred keys are obfuscation.

It was not my intention to offend you. Did you edit the file by hand? You _can_ do it, although it is not recommended, but you will have to play by calcurse rules.

You tacitly assume a lot. You cannot "unmap" keybindings. Your problems boil down to a wish of actions without any key bindings. Calcurse will use the keys configuration file you give it, but will always try to assign keys to actions that you have left undefined and warn you if that is the case. That is a perfectly reasonable behaviour in my opinion.

**Comment by lfos** [comment]
I agree that the current behavior is reasonable but I think the feature request to have actions without key bindings is reasonable too. Even though calcurse already warns about undefined key bindings, it currently lets you remove all bindings for an action and even save that state. Only if calcurse is started again, it will try to assign new bindings.

How do you feel about keeping the warning if the key configuration menu is left with undefined bindings but, when saving, create a special entry in the `keys` file with `undefined` as key binding, which is then ignored on followup calcurse launches?

**Comment by lhca** [comment]
@lfos:

> I agree that the current behavior is reasonable but I think the feature request to have actions without key bindings is reasonable too.

I don't. When I wrote

>  Calcurse will use the keys configuration file you give it, but will always try to assign keys to actions that you have left undefined and warn you if that is the case. 

it was not quite accurate. Calcurse wil _silently_ assign default keybindings to undefined actions and only warn you when a default is in use already. The `keys` file in question has 11 undefined actions:

```
generic-credits
generic-help
generic-import
generic-export
generic-prev-day
generic-next-day
generic-scroll-up
generic-scroll-down
start-of-week
end-of-week
pipe-item
```

Two warnings were issued with only a single action (pipe-item) left undefined. This raises the question which actions, and how many, can you meaningfully leave undefined for calcurse still to be usable?

@communistkiro wrote

> Given one can edit keybindings, ...

That's not a given. You can _configure_ the keybindings through the calcurse key configuration menu, but the configuration file `keys` is not a user interface as clearly stated in the file. This particular file has been maintained by hand, not by calcurse. That practice should be discouraged.

**Comment by communistkiro** [comment]
@lfos that'd be great. I was alternatively going to assign keys I don't use, say f10-f12 along with modifiers.

**Comment by lfos** [comment]
Lars, could you please elaborate more on why you think this is an unreasonable request? If it really is something we should not support at all, I think we should prevent the user from leaving the key bindings configuration with undefined bindings and not just display a warning. That being said, I don't see a reason to force users to assign random keys they don't use if they don't want to map a particular feature...

**Comment by lhca** [comment]
I wouldn't say its _un_reasonable, but neither is it reasonable. Things are
seldom black or white.

Before making changes I think we should make clear what we are changing, for
whom and how it should work. I see it in this way.

All virtual keys or actions shall have at least one keyboard key assigned. The
reason: simplicity. The user can at any time reconfigure the key mappings
through the keys configuration menu. Virtual keys that are sparingly or never
used can be mapped to sparingly used keyboard keys. In this respect calcurse is
very flexible.

Allowing virtual keys without any keybindings adds complexity without benefits.
What's the benefit of not being able to press a virtual key that you don't use
in the first place, and having "XXX" displayed for the keyboard key?

Currently calcurse will try to assign keyboard keys to _all_ virtual keys
(actions) on startup. The built-in default keys are used when no user-defined
key is defined. If a default key is already in use, a warning is issued and the
virtual key is left undefined unless another default key is available. The user
may leave a virtual key undefined by removing all the defined bindings in the
keys configuration menu; when the keys file is saved, the virtual key is left
out.

I agree that the current behaviour of allowing undefined bindings is
inconsistent with the initial description and should be tightened. Calcurse
should refuse to start if not all virtual keys are (or can be) mapped. This
happens already when more than five errors are encountered in processing the
keys file. Preventing the user from leaving the keys configuration menu with
undefined bindings is more problematic. Undefined keybindings will be resolved
on the next invocation of calcurse, possibly by calcurse refusing to start.

The keys file and user behaviour which lead to this discussion are unusual, I
would claim, and not a good starting point for a discussion of improvements for
the majority. In general, users are not forced to assign (random) keys; calcurse
does it. In this exceptional case the keys file was edited by hand and contained
virtual keys without any keybindings. Calcurse leaves out such virtual keys when
writing the file, and it is a small marvel that the file was accepted. Calcurse
should have rejected it.

**Comment by lfos** [comment]
When saying that simplicity is the reason, do you mean simplicity from a user's perspective, from a more holistic design perspective or from a code maintenance perspective?

From a user's perspective, is it really simple to have to think of a key that you would not use if you partically want to leave a binding undefined? Wouldn't we provide peace of mind by supporting just marking a feature as unused and then giving the user the option to change it later if they change their mind? You probably use all/most calcurse features, and so do I, but shouldn't we support "I think I will never need this particular feature and I don't want to spend a second thinking of a weird key to map this to?"

From a design perspective, since we already allow an arbitrary number of bindings, is special casing 0 a simplification?

From a coding perspective, would adding support for an `undef` keyword really be that much work, given that we should already support undefined bindings for the case that it cannot be mapped when starting calcurse?

**Comment by lhca** [comment]
On Sat, Jun 20, 2020 at 12:07:47PM -0700, Lukas Fleischer wrote:
>    When saying that simplicity is the reason, do you mean simplicity from a
>    user's perspective, from a more holistic design perspective or from a code
>    maintenance perspective?

I'm slowly coming round to your point of view. Maybe simplicity really meant
that since there is no bug, nothing has to be done;-)

In the keys file there are three possibilities for each action:

1. One or several keys are assigned to it (as now)
2. It is marked as unassigned or undefined (new)
3. It is missing from the file (as now)

On load of the keys file, calcurse respectively

1. Assigns the key(s)
2. Assigns "undefined" (or similar)
3. Assigns a default key if possible

After load all actions should either have keys assigned or be undefined. If not,
calcurse should exit with a failure. If there are any syntax/semantic errors,
calcurse should reject the file and exit.

If not all actions are provided for in the keys file, and default keys are
assigned, the user should be informed (this is not the case now). For example,
"Default keys have been assigned to 2 actions." For a new installation this
would be "Default keys have been assigned to 47 actions." Or should this case
be an exception?

The keys file is only saved on exit from the configuration menus, which means
that a message about default keys being assigned will be shown at start until
the user enters the configuration menus. This behaviour can be considered a
"nudge" to the user to acknowledge the assignment or amend it by entering the
keys configuration manu. Alternatively calcurse could also save the keys file on
exit. Or better: immediately, if default keys are assigned.

Presently, when a user leaves the key configuration menu, he is warned about
undefined actions. This should still be the case, but now this state of affairs
is also saved in the keys file.

**Comment by lfos** [comment]
Yes, it's definitely more of a feature request, not a bug. I agree with all your suggestions. The informative message for automatic assignments is optional but it's nice to have. If we decide to show it, I think it's fine to display it always, even on the first run.

**Comment by communistkiro** [comment]
Good job, guys.

**Comment by lfos** [comment]
Has this been fixed yet?

**Comment by lhca** [comment]
No.
